server/service/src/sync/integrate_document.rs

```diff
@@ -0,0 +1,87 @@
+use repository::{
+    Document, DocumentRegistryFilter, DocumentRegistryRepository, DocumentRepository, EqualFilter,
+    RepositoryError, StorageConnection,
+};
+
+use crate::programs::{
+    encounter::{
+        encounter_updated::update_encounter_row, validate_misc::validate_encounter_schema,
+    },
+    patient::{patient_schema::SchemaPatient, patient_updated::patient_document_updated},
+    program_enrolment::program_enrolment_updated::program_enrolment_updated,
+    program_enrolment::program_schema::SchemaProgramEnrolment,
+};
+
+pub(crate) fn upsert_document(
+    con: &StorageConnection,
+    document: &Document,
+) -> Result<(), RepositoryError> {
+    let registry = DocumentRegistryRepository::new(con)
+        .query_by_filter(
+            DocumentRegistryFilter::new().document_type(EqualFilter::equal_to(&document.r#type)),
+        )?
+        .pop()
+        .ok_or(RepositoryError::NotFound)?;
+
+    DocumentRepository::new(con).insert(document)?;
+
+    match registry.context {
+        repository::DocumentContext::Patient => update_patient(con, document)?,
```

Comment:
If there are multiple document for the same patient, does this mean we will be updating patient record each time ?

I thought your plan was to update the patient when document is opened ?

Does this PR account for circular sync ?

Reply:
Do we always sync documents in chronological order ?

Reply:
Yes the name table get updated every time when there is a new document version. It doesn't assumes chronological order for the incoming documents though.

> I thought your plan was to update the patient when document is opened ?

Yes its still the plan :)
https://github.com/openmsupply/open-msupply/issues/872

> Does this PR account for circular sync ?
is already handled in a later PR...

Reply:
> Yes the name table get updated every time when there is a new document version. It doesn't assumes chronological order for the incoming documents though.

So during initialisation for example, we can't assume that name would be in the latest state. Since we get the latest state of name (as it itself syncs) why should we update_patient from document during sync ?

Reply:
You mean before or after #872? Yes after #872 we don't need to update the name table here.

For now we probably do not need it either (apart from possible sync edge cases?)

Reply:
Yeah would only be an issue if there are multiple documents (multiple edits), and we re-sync the site, a couple of other edge cases too, but shouldn't affect one site editing one name and not re-syncing